Add option to regenerate the primary agent cert.

m0dular opened the pull request:
Prior to this commit, the configure_master step would fail if the
primary agent cert had also expired.  This commit adds a check for this
cert's expiration and an option to regenerate it.  This is done by
deleting the pem encoded cert and keys.

m0dular committed:
Add option to regenerate the primary agent cert.

Prior to this commit, the configure_master step would fail if the
primary agent cert had also expired.  This commit adds a check for this
cert's expiration and an option to regenerate it.  This is done by
deleting the pem encoded cert and keys.

m0dular committed:
Bump version

jarretlavallee commented on tasks/check_primary_cert.sh:
```diff
@@ -0,0 +1,18 @@
+expiry_date="${expiry_date#*=}"
+expiry_seconds="$(date --date="$expiry_date" +"%s")" || fail "Error calculating expiry date from enddate"
+
+if (( $(date +"%s") >= expiry_seconds )); then
```
How about an else case here that warns if it is going to expire in the next 3 months. Users could be running this proactively and it has yet to have expired but may not know about the regen option.

m0dular replied:
Added.

jarretlavallee commented on tasks/configure_master.sh:
```diff
@@ -1,12 +1,21 @@
+PUPPET_BIN='/opt/puppetlabs/puppet/bin'
 
-\cp -R /etc/puppetlabs/puppet/ssl /var/puppetlabs/backups || fail "Error backing up '/etc/puppetlabs/puppet/ssl'"
+mkdir -p /var/puppetlabs/backups/
```
We need to ensure security on the backup of the `ssl` directory. Can we do `cp -a` or chmod 700 on the dir, etc?

m0dular replied:
`chmod` is a good idea.

m0dular replied:
Oh I see, I thought you meant the `backups` directory itself.  We can do `cp -a` to preserve the `ssl` permissions.

m0dular replied:
Done.

jarretlavallee commented on plans/extend_ca_cert.pp:
```diff
@@ -2,6 +2,7 @@
   TargetSpec $targets,
   Optional[TargetSpec] $compile_masters = undef,
   $ssldir                               = '/etc/puppetlabs/puppet/ssl',
+  $regen_primary_cert                   = false,
```
It seems as though we should default this to `true` and make it clear we can disable it.

m0dular replied:
Are you ok with defaulting to `false` with the new error and warning messages?  I also added the param to the reference.

jarretlavallee commented on tasks/check_primary_cert.sh:
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+declare PT__installdir
+# shellcheck disable=SC1090
```
SC1090 is disabled in most files. Let's disable it globally in CI. `SHELLCHECK_OPTS="--exclude SC1090"` to reduce the redundant disabling of it. This is probably a separate PR, but shellcheck can be added to the travis-ci.yml like the following. 

https://github.com/puppetlabs/support-tasks/blob/master/.sync.yml#L25-L34
https://github.com/puppetlabs/support-tasks/blob/master/.sync.yml#L48-L49
https://github.com/puppetlabs/support-tasks/blob/master/.sync.yml#L59-L62

Then run `pdk update` to sync the changes to the travis file.

m0dular replied:
Are you ok with opening a new issue for setting up and configuring `shellcheck`?

jarretlavallee replied:
Yup

jarretlavallee reviewed (changes requested):
It looks good to me. I have a few thoughts for discussion, but nothing that would be a blocker.

The `REFERENCE.md` needs to be updated manually until we fix #4

jarretlavallee commented on tasks/check_primary_cert.sh:
```diff
@@ -0,0 +1,18 @@
+source "$PT__installdir/ca_extend/files/common.sh"
+PUPPET_BIN='/opt/puppetlabs/puppet/bin'
+
+expiry_date="$($PUPPET_BIN/openssl x509 -enddate -noout -in "$($PUPPET_BIN/puppet config print hostcert)")" || {
```
This could be a failure point. If they have already tried to run the plan with a regeneration of the primary's cert, it would have been deleted. There is no way to rerun the plan in this case. We should allow for moving forward if this check fails.

m0dular replied:
This task is not run if `$regen_primary_cert` is `true`.

jarretlavallee replied:
If someone runs it with `regen_primary_cert = false` and the cert has already been cleaned it will fail with no clear method of moving forward. The failure could point to running it with `regen_primary_cert = true` to get around the failure.

m0dular replied:
Ah, that's not a scenario I thought of.  I'll add a check for the file and a useful error message.

m0dular replied:
Also added in `2a3899a`

m0dular committed:
Add a check for a soon-to-expire primary cert

Prior to this commit, the task would only check for an already expired
cert.  This commit checks for whether it will expire within 3 months and
produces a warning.  It also references the flag to regen the cert.

m0dular committed:
Preserve permissions when backing up ssldir

m0dular committed:
Add regen_primary_cert flag to REFERENCE.md

m0dular committed:
Rubocop fix

m0dular commented:
Not totally sure how this works, but it's still showing the `REFERENCE.md` part as a requested change.  I added the parameter there, with the larger goal of reworking it in #4.  Are there any other items to address?

jarretlavallee reviewed (approved):
👍 I haven't tested it but it looks good to me

m0dular commented:
Cool, I tested the different scenarios by using a date farther into the future so it thought my cert were expired.  One of my goals is to have full test coverage, so we'll get there eventually.
